Subject: mysys: portable CRC-32/CRC-32C on OpenVMS x86-64

The x86 CRC code needs the compiler's intrinsic headers (cpuid.h,
emmintrin.h, nmmintrin.h, ...). VSI C++ (clang 10) has the builtins but
ships none of those headers, so build the portable implementations on
OpenVMS. Slower checksums until the SIMD code can be enabled (see
docs/DECISIONS.md).

--- a/mysys/CMakeLists.txt	2026-10-04 13:22:32.800899809 +0100
+++ b/mysys/CMakeLists.txt	2026-10-04 13:22:32.852026400 +0100
@@ -76,7 +76,8 @@
       COMPILE_FLAGS "-march=armv8-a+crc+crypto"
   )
   ENDIF()
-ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64|amd64|i386|i686")
+# OpenVMS: VSI C++ ships no x86 intrinsic headers; use the portable code.
+ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64|amd64|i386|i686" AND NOT VMS)
   SET(MYSYS_SOURCES ${MYSYS_SOURCES} crc32/crc32_x86.c crc32/crc32c_x86.cc)
   IF(CMAKE_C_COMPILER_ID STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_LESS "5")
     SET_SOURCE_FILES_PROPERTIES(crc32/crc32_x86.c PROPERTIES
--- a/mysys/crc32ieee.cc	2026-10-04 13:22:32.817900017 +0100
+++ b/mysys/crc32ieee.cc	2026-10-04 13:22:32.852200975 +0100
@@ -28,7 +28,8 @@
 
 typedef unsigned int (*my_crc32_t)(unsigned int, const void *, size_t);
 
-#if defined _M_IX86 || defined _M_X64 || defined __i386__ || defined __x86_64__
+#if (defined _M_IX86 || defined _M_X64 || defined __i386__ || defined __x86_64__) && \
+  !defined __VMS /* no x86 intrinsic headers in VSI C++ */
 extern "C" my_crc32_t crc32_pclmul_enabled();
 #elif defined HAVE_ARMV8_CRC
 extern "C" int crc32_aarch64_available();
@@ -38,7 +39,8 @@
 
 static my_crc32_t init_crc32()
 {
-#if defined _M_IX86 || defined _M_X64 || defined __i386__ || defined __x86_64__
+#if (defined _M_IX86 || defined _M_X64 || defined __i386__ || defined __x86_64__) && \
+  !defined __VMS /* no x86 intrinsic headers in VSI C++ */
   if (my_crc32_t crc= crc32_pclmul_enabled())
     return crc;
 #elif defined HAVE_ARMV8_CRC
--- a/mysys/crc32/crc32c.cc	2026-10-04 13:22:32.834229454 +0100
+++ b/mysys/crc32/crc32c.cc	2026-10-04 13:22:32.852909557 +0100
@@ -349,7 +349,8 @@
 # include <intrin.h>
 # include <immintrin.h>
 # define USE_SSE42 /* nothing */
-#elif defined __GNUC__ && (defined __i386__||defined __x86_64__)
+#elif defined __GNUC__ && (defined __i386__||defined __x86_64__) && \
+  !defined __VMS /* no x86 intrinsic headers in VSI C++ */
 # if __GNUC__ < 5 && !defined __clang_major__
 /* the headers do not really work in GCC before version 5 */
 #  define _mm_crc32_u8(crc,data) __builtin_ia32_crc32qi(crc,data)
@@ -489,7 +490,8 @@
 #if defined(HAVE_ARMV8_CRC)
 extern "C" my_crc32_t crc32c_aarch64_available(void);
 extern "C" const char *crc32c_aarch64_impl(my_crc32_t);
-#elif defined __i386__||defined __x86_64__||defined _M_X64||defined _M_IX86
+#elif (defined __i386__||defined __x86_64__||defined _M_X64||defined _M_IX86) && \
+  !defined __VMS
 extern "C" my_crc32_t crc32c_x86_available(void);
 extern "C" const char *crc32c_x86_impl(my_crc32_t);
 #endif
@@ -502,7 +504,8 @@
 #elif defined HAVE_ARMV8_CRC
   if (my_crc32_t crc= crc32c_aarch64_available())
     return crc;
-#elif defined __i386__||defined __x86_64__||defined _M_X64||defined _M_IX86
+#elif (defined __i386__||defined __x86_64__||defined _M_X64||defined _M_IX86) && \
+  !defined __VMS
   if (my_crc32_t crc= crc32c_x86_available())
     return crc;
 #endif
@@ -519,7 +522,8 @@
 #elif defined HAVE_ARMV8_CRC
   if (const char *ret= crc32c_aarch64_impl(ChosenExtend))
     return ret;
-#elif defined __i386__||defined __x86_64__||defined _M_X64||defined _M_IX86
+#elif (defined __i386__||defined __x86_64__||defined _M_X64||defined _M_IX86) && \
+  !defined __VMS
   if (const char *ret= crc32c_x86_impl(ChosenExtend))
     return ret;
 #endif
